Add ADOT Ruby Documentation (#224)
aws-otel/aws-otel.github.io: src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx

```ruby
gem 'opentelemetry-exporter-otlp'
gem 'opentelemetry-sdk'

gem 'opentelemetry-propagator-xray'
```

Or, install them directly:

```bash
$ gem install opentelemetry-exporter-otlp \
              opentelemetry-sdk \
              opentelemetry-propagator-xray
```

Next, we'll use `bundler` to install gems that automatically instrument your application code.

OpenTelemetry Ruby distributes many gems that instrument well-known Ruby dependencies. You need to install the relevant instrumentation package for every dependency you want to generate traces for. To see supported gems, check out the [OpenTelemetry Registry](https://opentelemetry.io/registry/?s=ruby&component=instrumentation).

For example, use `bundler` and add the follow instrumentation gems to your Gemfile:

```ruby
gem 'opentelemetry-instrumentation-faraday', '~> 0.19'
gem 'opentelemetry-instrumentation-rails', '~> 0.20'
```

Or, install them directly:

```bash
$ gem install opentelemetry-instrumentation-faraday -v '~> 0.19' \
              opentelemetry-instrumentation-rails -v '~> 0.20'
```

<SectionSeparator />

> ## Setting up the Global Tracer

Reviewer: Sorry for missing this initially - I added a new section for configuring sampling since that's been a pain point for existing customers: https://github.com/willarmiros/aws-otel-docs/blob/main/templates/manual-trace-instrumentation.md#configuring-sampling

If we can let's include this sampling configuration too.

Author: Cool, I investigated what it would take to configure it with environment variables and with code and added it as a section here!

### Sending Traces to AWS X-Ray

> For a ruby on rails application, OpenTelemetry Ruby [recommends using a Rails initializer](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) to configure the OpenTelemetry Ruby SDK. ADOT provides a [working example of such an intializer in our sample app repo](https://github.com/aws-observability/aws-otel-ruby/blob/main/sample-apps/manual-instrumentation/ruby-on-rails/config/initializers/opentelemetry.rb). A [railtie](https://api.rubyonrails.org/classes/Rails/Railtie.html) could also be used for this purpose.

Author: I am also unfamiliar with Railtie, but it has come up when I ask OTel Ruby SIG members question and I noticed a SIG member mentioned it to me here: https://github.com/open-telemetry/opentelemetry-ruby/issues/1089#issuecomment-1015914979

There's less evidence than I thought originally to support it being a canonical thing in Ruby, so I will remove it! Especially since we don't have any concrete examples.

In a Ruby on Rails app, you will not need to require packages in your application code because of [autoloading](https://guides.rubyonrails.org/getting_started.html#autoloading). This assumes you are using `bundler` and a Gemfile.

> However, if you tracing a normal Ruby program, or have included gems with the `, require: false` option, the code below shows step by step how to configure the OpenTelemetry Ruby SDK in any normal Ruby program.

Author: Sure I can split it up. For reference, [the OTel Ruby docs](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) talk about Rails in their main documentation section which signal high importance to me. But I'll split it up regardless

First, you will need to "require" the gems you use for your application and the ones used by OpenTelemetry manually. 

```ruby lineNumbers=true
# Basic packages for your application
require 'aws-sdk'
require 'faraday'

# Add imports for OTel components into the application
require 'opentelemetry-api'
require 'opentelemetry-exporter-otlp'
require 'opentelemetry-sdk'

# Import the gem containing the AWS X-Ray for OTel Ruby ID Generator and propagator
require 'opentelemetry-propagator-xray'
```

> Next, configure the Global Tracer Provider. This involves configuring

Reviewer: Wow, this section has a lot of detail! I think as a general rule of thumb for documentation, we can assume 2 things: 
1. Most customers do not want to read a ton, they just want to get stuff working
2. Some customers will want to get in the weeds and understand very specific things, and those will be the ones who deeply read the docs.

With that in mind, I think we can simplify this section by first having a brief description of what the customer should do, then put the large code snippet that "puts everything together" first with comments that also explain each line. Then we can remove the individual code snippets. The additional details, like environment variable configuration and TLS stuff, can go after the main code snippet for those power users.

Author: Done!

span_processor = Trace::Export::BatchSpanProcessor.new(otlp_exporter)
```

The `endpoint=` argument allows you to set the address that the exporter will use to connect to the collector. If unset, the SDK will try to connect to `http://localhost:4318` by default. Note that because the scheme is `http` by default, you have to explicitly set it to be `https` if necessary.

If the Collector the application will connect to is running without TLS configured, the `http` scheme is used to disable client transport security for the OTLP exporter's connection. This option should never be used in production, non-sidecar deployments.

If the Collector the application will connect to is running with TLS configured, the `https` scheme and the `certificate_file=/path/to/cert.pem` argument should be used to give a path to credentials that allow the application to establish a secure connection for the app's exporter. The credentials at this path should be the public certificate of the collector, or one of its root certificates.

With the exporter created, initialize OpenTelemetry Ruby SDK to use it:

```ruby
OpenTelemetry::SDK.configure do |c|
  c.span_processors = [span_processor]
end

> ```

Reviewer: Let's remove these individual code snippets in this section.

Author: Removed!

Next, because the AWS X-Ray backend service requires compliant Trace IDs, create an instance of the ID generator during initialization as follows:

```ruby
OpenTelemetry::SDK.configure do |c|
  c.id_generator = OpenTelemetry::Propagator::XRay::IDGenerator
end
```

To allow the span context to propagate downstream when the application makes calls to external services, configure the global propagator to use the AWS X-Ray Propagator.

The `OTEL_PROPAGATORS` environment variable can be configured to have the OpenTelemetry Ruby SDK automatically find and initialize the propagator.

```
OTEL_PROPAGATORS=xray
```

On the other hand, you can set the global propagator in code as follows.

```ruby
OpenTelemetry::SDK.configure do |c|
  c.propagators = [OpenTelemetry::Propagator::XRay::TextMapPropagator.new]
end
```

Either way, the propagator should be configured as soon as possible in your application's code.

Putting this all together we have something like the following:

```ruby

> OpenTelemetry::SDK.configure do |c|
>   c.service_name = 'aws-otel-manual-rails-sample'
> 
>   c.span_processors = [
>       Trace::Export::BatchSpanProcessor.new(
>           OpenTelemetry::Exporter::OTLP::Exporter.new(endpoint="http://localhost:4318")
>       )
>     ]
>   c.id_generator = OpenTelemetry::Propagator::XRay::IDGenerator
>   c.propagators = [OpenTelemetry::Propagator::XRay::TextMapPropagator.new]
> end

Reviewer: ```suggestion
OpenTelemetry::SDK.configure do |c|
  c.service_name = 'aws-otel-manual-rails-sample'

  # Comment explaining batch span processor
  c.span_processors = [
      Trace::Export::BatchSpanProcessor.new(
          OpenTelemetry::Exporter::OTLP::Exporter.new(endpoint="http://localhost:4318")
      )
    ]
    
  # Comment explaining Id generator
  c.id_generator = OpenTelemetry::Propagator::XRay::IDGenerator
  
  # Comment on propagator
  c.propagators = [OpenTelemetry::Propagator::XRay::TextMapPropagator.new]
end
```

Instead, we can add some comments in the larger code snippet to explain.

Author: Added comments.

```

> ### Using the AWS resource Detectors

Author: Removed!

gem 'opentelemetry-instrumentation-aws_sdk', '~> 0.2.1'
```

Otherwise you can install it directly.

```bash
gem install opentelemetry-instrumentation-aws_sdk -v '~> 0.2.1'
```

**NOTE:** Since these instrumentations are not yet stable, we recommend installing it at a pinned version.

Instrumenting the AWS SDK is as easy as initializing the OpenTelemetry Ruby SDK Configurator to use the `OpenTelemetry::Instrumentation::AwsSdk` `Instrumentation` class. This initialization is done as soon as possible in your application so that subsequent calls using the SDK are wrapped by OpenTelemetry. This give OpenTelemetry the chance to record relevant information used by the SDK at the time of your application's call and export the information as spans.

Configuring it explicitly looks can look like the following. We set `suppress_internal_instrumentation` to `true` because we want calls that go into the AWS SDK to be terminal requests without tracing underlying HTTP calls and other things which would make the trace noise-y.

```ruby lineNumbers=true
OpenTelemetry::SDK.configure do |c|
  c.use 'OpenTelemetry::Instrumentation::AwsSdk', {
    suppress_internal_instrumentation: true
  }
end
```

> 
> Alternatively, once you've downloaded the `Instrumentation` gem, you can simply tell the OpenTelemetry SDK to trace all `Instrumentation`s including the `OpenTelemetry::Instrumentation::AwsSdk` Instrumentation.
> 
> ```ruby lineNumbers=true
> OpenTelemetry::SDK.configure do |c|
>   c.use_all()
> end
> ```

Reviewer: ```suggestion
```

Since this isn't specifically related to the AWS SDK, and it's mentioned above, let's remove.

For more information refer to [the upstream documentation for OpenTelemetry Ruby AWS SDK Instrumentation](https://github.com/open-telemetry/opentelemetry-ruby/tree/main/instrumentation/aws_sdk).

<SectionSeparator />

## Custom Instrumentation

### Creating Custom Spans

You can use custom spans to monitor the performance of internal activities that are not captured by instrumentation libraries. Note that only spans of kind `Server` are converted into X-Ray segments, all other spans are converted into X-Ray subsegments. For more on segments and subsegments, see the [AWS X-Ray developer guide](https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-segments).

```ruby lineNumbers=true
require 'aws-sdk'
require 'opentelemetry-api'

# Get a tracer from the Global Tracer Provider
tracer = OpenTelemetry.tracer_provider.tracer('my-tracer')

> tracer.in_span('Root Span') do |root_span|

Author: Updated! Although on a side note, I thought we changed that last year so that it didn't need to be of type `Server`? Either way I have updated it.

path: '/docs/getting-started/ruby-sdk/trace-manual-instr'
---

import SectionSeparator from "components/MdxSectionSeparator/sectionSeparator.jsx"
import SubSectionSeparator from "components/MdxSubSectionSeparator/subsectionSeparator.jsx"

## Introduction

With OpenTelemetry Ruby manual instrumentation, you configure the OpenTelemetry SDK within your application with just a few lines of code. OpenTelemetry Ruby then automatically produces trace spans with telemetry data describing the values used by the Ruby gems in your application. This telemetry data can then be exported to a backend like AWS X-Ray using the `OpenTelemetry::Propagator::XRay::IDGenerator` found in the ADOT Ruby `opentelemetry-propagator-xray` gem. We also strongly recommend using the `OpenTelemetry::Propagator::XRay::TextMapPropagator` propagator found in the same gem to support propagating the trace context across AWS services. This propagator handles the extraction and injecting of the [AWS X-Ray Tracing header](https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-tracingheader) for requests from or to remote services.

> In this guide, we walkthrough the steps needed to trace an application with manual instrumentation.

Author: We should use 2 words for this part here

```suggestion
In this guide, we walk through the steps needed to trace an application with manual instrumentation.
```

```

Or, install them directly:

```bash
$ gem install opentelemetry-instrumentation-faraday -v '~> 0.19' \
              opentelemetry-instrumentation-rails -v '~> 0.20'
```

<SectionSeparator />

## Setting up the Global Tracer

### Sending Traces to AWS X-Ray

Manual Instrumentation with OpenTelemetry Ruby involves configuring the OpenTelemetry Ruby SDK. Below we discuss where configuration should take place based on the Ruby program you are instrumenting.

#### From a Rails Application

For a ruby on rails application, OpenTelemetry Ruby [recommends using a Rails initializer](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) to configure the OpenTelemetry Ruby SDK. ADOT provides a [working example of such an intializer in our sample app repo](https://github.com/aws-observability/aws-otel-ruby/blob/main/sample-apps/manual-instrumentation/ruby-on-rails/config/initializers/opentelemetry.rb).

In a Ruby on Rails app, you will not need to require packages in your application code because of [autoloading](https://guides.rubyonrails.org/getting_started.html#autoloading). This assumes you are using `bundler` and a Gemfile.

> #### From a normal Ruby Program

Reviewer: ```suggestion
#### From other Ruby Programs
```

This might confuse Rails users, who probably consider their programs "normal" ;)

Author: Lol, I must be dizzy or something, I could've sworn "normal" is what the OTel Ruby docs said, but [they say "other" too](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) so thanks for htis!

This section describes recommended configuration needed for OpenTelemetry X-Ray tracing with X-Ray.

> ##### Quick Setup

Reviewer: Should be ok

> When you instrument a normal Ruby program, or if you included the gems with the `require: false` option, you will need to "require" the gems distributed by OpenTelemetry manually.

Reviewer: ```suggestion
When you instrument a non-rails Ruby program, or if you included the gems with the `require: false` option, you will need to "require" the gems distributed by OpenTelemetry manually.
```

require 'opentelemetry-api'
require 'opentelemetry-exporter-otlp'
require 'opentelemetry-sdk'

# Import the gem containing the AWS X-Ray for OTel Ruby ID Generator and propagator
require 'opentelemetry-propagator-xray'

# Configure OpenTelmetry Ruby SDK
OpenTelemetry::SDK.configure do |c|
  # Set the service name to identify your application in the X-Ray backend service map
  c.service_name = 'aws-otel-manual-rails-sample'

  c.span_processors = [
      # Use the BatchSpanProcessor to send traces in groups instead of one at a time
      Trace::Export::BatchSpanProcessor.new(
          # Use the default OLTP Exporter to send traces to the ADOT Collector
          OpenTelemetry::Exporter::OTLP::Exporter.new(
            # The ADOT Collector is running as a sidecar and listening on port 4318
            endpoint="http://localhost:4318"
          )
      )
    ]

  # The X-Ray ID Generator generates spans with X-Ray backend compliant IDs
  c.id_generator = OpenTelemetry::Propagator::XRay::IDGenerator

  # The X-Ray Propagator injects the X-Ray Tracing Header into downstream calls
  c.propagators = [OpenTelemetry::Propagator::XRay::TextMapPropagator.new]
end
```

> With this, your Ruby application has configured OpenTelemetry Ruby to trace for the X-Ray backend service! To automatically trace popular Ruby gems, jump to the next section to learn about [instrumenting with OpenTelemetry Ruby Instrumentations](#instrumenting-an-application).

Reviewer: Yep it should!

##### Advanced Setup

From above, we learned that configuring OpenTelemetry Ruby required specifying 3 core steps
* which Exporter to use to export to the [ADOT Collector](https://aws-otel.github.io/docs/getting-started/collector)
* which ID Generator to use to generate Trace IDs
* which Propagator to use to propagate Trace Context to downstream calls

By default, OpenTelemetry Ruby SDK [is already configured to initialize an OTLP exporter](https://github.com/open-telemetry/opentelemetry-ruby/tree/main/exporter/otlp#how-do-i-get-started). The exporter can also be completely [configured using environment variables](https://github.com/open-telemetry/opentelemetry-ruby/tree/main/exporter/otlp#how-can-i-configure-the-otlp-exporter).

```bash

> OTEL_TRACES_EXPORTER=otlp && OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318

Reviewer: ```suggestion
export OTEL_TRACES_EXPORTER=otlp
export OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
```

```

The `OTEL_EXPORTER_OTLP_ENDPOINT` value allows you to set the address that the exporter will use to connect to the collector. If unset, the SDK will try to connect to `http://localhost:4318` by default. Note that because the scheme is `http` by default, you have to explicitly set it to be `https` if necessary.

If the Collector the application will connect to is running without TLS configured, the `http` scheme is used to disable client transport security for the OTLP exporter's connection. This option should never be used in production, non-sidecar deployments.

If the Collector the application will connect to is running with TLS configured, the `https` scheme and the `certificate_file=/path/to/cert.pem` argument should be used to give a path to credentials that allow the application to establish a secure connection for the app's exporter. The credentials at this path should be the public certificate of the collector, or one of its root certificates.

Next, because the AWS X-Ray ID Generator can only be configured through code, you cannot use an environment variable to select it. The ID Generator must be used at the time OpenTelemetry Ruby SDK is configured.

> Finally,tTo allow the span context to propagate downstream when the application makes calls to external services, configure the global propagator to use the AWS X-Ray Propagator.

Reviewer: ```suggestion
Finally, to allow the span context to propagate downstream when the application makes calls to external services, configure the global propagator to use the AWS X-Ray Propagator.
```

The `OTEL_PROPAGATORS` environment variable can be configured to have the OpenTelemetry Ruby SDK automatically find and initialize the propagator.

```
OTEL_PROPAGATORS=xray
```

The propagator should be configured as soon as possible in your application's code so that subsequent downstream requests get the OpenTelemetry trace context injected into its HTTP headers. This is what allows your traces to be connected and for you to see a complete Service Graph in the X-Ray console.

Likewise, configuring the X-Ray Propagator means incoming requests to your application can parse out an OpenTelemetry Trace context and use the same Trace ID to pick up tracing where the upstream service left off.

You can combine the `xray` propagator with other propagators like `tracecontext` and `b3` just fine, but it is recommended you put `xray` last because the propagator later in the list will override previous propagators.

### Configuring Sampling

By default, the OpenTelemetry Ruby SDK follows the parent span's sampling decision if it exists, and samples 100% of incoming requests otherwise. This is known as the `parentbased_always_on` sampler.

> To reduce the sampling rate, configure OpenTelemetry Ruby SDK to use the `parentbased_traceidratio` sampler. This can be configured using [the OpenTelemetry Specification defined environment variables](https://github.com/open-telemetry/opentelemetry-specification/blob/7b504383f53b02b10f62ef78fa008fdfd18c633e/specification/sdk-environment-variables.md#general-sdk-configuration):

Reviewer: ```suggestion
To reduce the sampling rate, configure OpenTelemetry Ruby SDK to use the `parentbased_traceidratio` sampler. This can be configured using [the OpenTelemetry Specification defined environment variables](https://github.com/open-telemetry/opentelemetry-specification/blob/7b504383f53b02b10f62ef78fa008fdfd18c633e/specification/sdk-environment-variables.md#general-sdk-configuration). For instance, to reduce the sampling rate to 10% of requests, set the following environment variables:
```

```bash

> OTEL_TRACES_SAMPLER=traceidratio && OTEL_TRACES_SAMPLER_ARG=0.10

Reviewer: ```suggestion
export OTEL_TRACES_SAMPLER=traceidratio
export OTEL_TRACES_SAMPLER_ARG=0.10
```

Reviewer: Also, just want to confirm this is the `parentbased` sampler right?

Author: Oh sorry no it is not! I will fix that right away.

Author: ```suggestion
OTEL_TRACES_SAMPLER=parentbased_traceidratio && OTEL_TRACES_SAMPLER_ARG=0.10
```

```

Alternatively, this can be set on the global TracerProvider after the OpenTelemetry Ruby SDK has been configured:

```ruby

> OpenTelemetry.tracer_provider.sampler = Samplers.trace_id_ratio_based(0.10)

Reviewer: Is this `parentbased`?

Author: Ah no it is not, will update right away.

Author: ```suggestion
OpenTelemetry.tracer_provider.sampler = Samplers.parent_based(root: Samplers.trace_id_ratio_based(0.10))
```

```ruby lineNumbers=true
OpenTelemetry::SDK.configure do |c|
  c.use 'OpenTelemetry::Instrumentation::AwsSdk', {
    suppress_internal_instrumentation: true
  }
end
```

For more information refer to [the upstream documentation for OpenTelemetry Ruby AWS SDK Instrumentation](https://github.com/open-telemetry/opentelemetry-ruby/tree/main/instrumentation/aws_sdk).

<SectionSeparator />

## Custom Instrumentation

### Creating Custom Spans

You can use custom spans to monitor the performance of internal activities that are not captured by instrumentation libraries. Note that only spans of kind `Server` are converted into X-Ray segments, all other spans are converted into X-Ray subsegments. For more on segments and subsegments, see the [AWS X-Ray developer guide](https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-segments).

```ruby lineNumbers=true
require 'aws-sdk'
require 'opentelemetry-api'

# Get a tracer from the Global Tracer Provider
tracer = OpenTelemetry.tracer_provider.tracer('my-tracer')

tracer.in_span('Root Span', kind: :server) do |root_span|

>     p 'Started a root span'

Reviewer: ```suggestion
    p 'Started a root span, this will be a segment in the X-Ray console'
```

    tracer.in_span('Child Span') do |child_span|

>         p 'Started a child span'

Reviewer: ```suggestion
        p 'Started a child span, this will be a subsegment in the X-Ray console'
```

### Sending Traces to AWS X-Ray

Manual Instrumentation with OpenTelemetry Ruby involves configuring the OpenTelemetry Ruby SDK. Below we discuss where configuration should take place based on the Ruby program you are instrumenting.

#### From a Rails Application

For a ruby on rails application, OpenTelemetry Ruby [recommends using a Rails initializer](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) to configure the OpenTelemetry Ruby SDK. ADOT provides a [working example of such an intializer in our sample app repo](https://github.com/aws-observability/aws-otel-ruby/blob/main/sample-apps/manual-instrumentation/ruby-on-rails/config/initializers/opentelemetry.rb).

In a Ruby on Rails app, you will not need to require packages in your application code because of [autoloading](https://guides.rubyonrails.org/getting_started.html#autoloading). This assumes you are using `bundler` and a Gemfile.

#### From a normal Ruby Program

This section describes recommended configuration needed for OpenTelemetry X-Ray tracing with X-Ray.

> ##### Quick Setup

Reviewer: Just want to confirm - does the "quick setup" and "advanced setup" sections only apply to non-rails applications? If not, they shouldn't be nested like this.

Author: "Quick Setup" and "Advanced Setup" applies to _all_ applications, not just Rails applications.

Hm this is how I figured to split the section into "Rails" vs "non-Rails" as we mentioned previously. Would this split seem better?:
* Setting up the Global Tracer
  * Sending Traces to AWS X-Ray
    * Configure the OpenTelemetry Ruby SDK
    * Tips for Rails Applications

That's why I originally put "rails" in the "configure OTel Ruby SDK" section, because you still have to configure OTel Ruby, the bit about "Rails Initializers" is just a helpful tip. I guess this is [why OTel Ruby Getting Started Docs do not split Rails into its own section](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) but also just keep it together in the "General Initialization Section".

Let me know what you think? I think the "additional note" is a good compromise. Thank you!

Reviewer: Discussed offline, seems like we shouldn't have separate sections for rails vs. non-rails. We can structure it like:

* Setting up the Global Tracer
  * Sending Traces to AWS X-Ray (Includes a quick note about using rails initializers and has the "quick setup" content)
  * Configure the OpenTelemetry Ruby SDK (has the advanced setup content)

# Get a tracer from the Global Tracer Provider
tracer = OpenTelemetry.tracer_provider.tracer('my-tracer')

> tracer.in_span('Root Span', kind: :server) do |root_span|

Reviewer: Is the space between colons intentional? If so, man is Ruby weird

Author: Lol yes it is, you can see [an example on the OTel Ruby Getting Starter Docs - Manually Creating your own spans](https://opentelemetry.io/docs/instrumentation/ruby/manual_instrumentation/#creating-new-spans)
